server/src/main/java/org/elasticsearch/action/admin/cluster/stats/RemoteClusterActionListener.java

```diff
@@ -0,0 +1,68 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+package org.elasticsearch.action.admin.cluster.stats;
+
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.action.DelegatingActionListener;
+import org.elasticsearch.common.util.concurrent.CountDown;
+
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicReference;
+
+/**
+ * Action listener for operations that are performed on a group of remote clusters.
+ * It will wait for all operations to complete and then delegate to the upstream listener.
+ * Does not fail if one of the operations fails.
+ * <br>
+ * Returns a map of the results per cluster name via {@link #remoteListener(String)} method.
+ * This is the listener that should be used to perform the individual operation on the remote cluster.
+ *
+ * @param <T> the type of the individual per-cluster result
+ */
+public class RemoteClusterActionListener<T> extends DelegatingActionListener<T, Map<String, T>> {
```

Comment:
Could we just use a `RefCountingRunnable` rather than introducing a whole separate mechanism just for this one case?

Reply:
In fact you probably want a `org.elasticsearch.action.support.CancellableFanOut` here (and you'll need to set the parent task ID on all the remote-cluster requests to get cancellation to work properly too)

Reply:
OK I'll look into converting this to `CancellableFanOut`

Reply:
Looks like using `CancellableFanOut` simplifies it quite a bit, thanks for the advice.